### PR TITLE
feat(query-generator): ability to use indexHints option for mysql

### PR DESCRIPTION
**Pull request by jonmagic**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Have you added new tests to prevent regressions?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/master/CONTRIBUTING.md)?

### Description of change

Per https://github.com/sequelize/sequelize/issues/9421 this takes the proposed `indexHints` API and implements it for mysql.

```js
Project.findAll({
  indexHints: [
    { type: IndexHints.USE, values: ['index_project_on_name'] }
  ],
  where: {
    id: {
      [Op.gt]: 623
    },
    name: {
      [Op.like]: 'Foo %'
    }
  }
})
```

Will generate a mysql query that looks like this:

```sql
SELECT * FROM Project USE INDEX (index_project_on_name) WHERE name LIKE 'FOO %' AND id > 623;
```

### CODE REVIEW

File: lib/index-hints.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+'use strict';
+
+/**
+ * An enum of index hints to be used in mysql for querying with index hints
+ *
+ * @property USE
+ * @property FORCE
+ * @property IGNORE
+ */
+const IndexHints = module.exports = { // eslint-disable-line
```
Comment (SimonSchick) [inline lib/index-hints.js]: Please use `exports.IndexHints = { ... }`

File: lib/dialects/mysql/query-generator.js [hunk 2]
```diff
@@ -434,6 +435,24 @@ class MySQLQueryGenerator extends AbstractQueryGenerator {
     return result;
   }
 
+  selectFromTableFragment(options, model, attributes, tables, mainTableAs) {
+    let fragment = `SELECT ${attributes.join(', ')} FROM ${tables}`;
```
Comment (SimonSchick) [inline lib/dialects/mysql/query-generator.js]: Just to sanity check, do we need to escape anything here?

Comment (jonmagic) [reply lib/dialects/mysql/query-generator.js]: This is taken directly from abstract/query-generator.js so I don't believe so?

File: lib/index-hints.js (hunk 1 above)
Comment (jonmagic) [reply lib/index-hints.js]: Updated in 37772bcb0c0a7bc9eb1982466ce0f9a4dbe341f0

File: lib/dialects/mysql/query-generator.js [hunk 3]
```diff
@@ -434,6 +435,24 @@ class MySQLQueryGenerator extends AbstractQueryGenerator {
     return result;
   }
 
+  selectFromTableFragment(options, model, attributes, tables, mainTableAs) {
```
Comment (sushantdhiman) [inline lib/dialects/mysql/query-generator.js]: No need to override this you can implement this with `dialect.supports`.

1. First start with MySQL, define `indexHints: true` someone where [here](https://github.com/sequelize/sequelize/blob/2c18e084509469c7e8b9fdbabcb4c713312c216d/lib/dialects/mysql/index.js#L43)
2. Along with `options.indexHints` you will now also check for `this._dialect.supports.indexHints`

Comment (jonmagic) [reply lib/dialects/mysql/query-generator.js]: Done in bbd3c2d5dd5aced953bb6b50ca00f2ec8191ba7c

File: lib/dialects/abstract/query-generator.js [hunk 4]
```diff
@@ -1906,6 +1907,14 @@ class QueryGenerator {
       fragment += ` AS ${mainTableAs}`;
     }
 
+    if (options.indexHints && this._dialect.supports.indexHints) {
+      for (const hint of options.indexHints) {
+        if (IndexHints[hint.type]) {
+          fragment += ` ${IndexHints[hint.type]} INDEX (${hint.values.join(',')})`;
```
Comment (sushantdhiman) [inline lib/dialects/abstract/query-generator.js]: You will need to quote index names from `values`

```js
hint.values.map(indexName => this.quoteIdentifiers(indexName)).join(',')
```

Comment (jonmagic) [reply lib/dialects/abstract/query-generator.js]: Updated in 65c91b178d1f5427489ac458298cd2557d2c135e

**Pull request merged**